File: src/claude_usage_monitor/poller.py
"""Background polling thread.

Runs fetch_all() on a configurable interval, calls registered callbacks
with the fresh UsageData or an error string. Thread-safe via threading.Event.
"""

from __future__ import annotations

import logging
import threading
from typing import Callable

from .client import (
    ClaudeClientError,
    RateLimitedError,
    SessionExpiredError,
    fetch_all,
)
from .config import Config
from .firefox_cookies import (
    CookieError,
    FirefoxCookieError,
    build_cookie_header,
    extract_org_id,
    get_claude_cookies,
)
from .models import UsageData

logger = logging.getLogger(__name__)

OnDataCallback = Callable[[UsageData], None]
OnErrorCallback = Callable[[str], None]


class Poller:
    """Periodic background worker that fetches claude.ai usage data."""

    def __init__(
        self,
        config: Config,
        on_data: OnDataCallback,
        on_error: OnErrorCallback,
    ) -> None:
        self._config = config
        self._on_data = on_data
        self._on_error = on_error
        self._stop_event = threading.Event()
        self._thread = threading.Thread(target=self._loop, daemon=True, name="poller")
        self._force_event = threading.Event()
        # Doubles after each consecutive 429, resets to 1 on success.
        # A manual "Refresh now" still bypasses the wait via _force_event.
        self._backoff_factor = 1

    def start(self) -> None:
        logger.info("Poller starting (interval=%ds)", self._config.poll_interval_seconds)
        self._thread.start()

    def stop(self) -> None:
        logger.info("Poller stopping.")
        self._stop_event.set()
        self._force_event.set()
        # stop() can run on the Tk thread (quit button): keep the join short —
        # an in-flight HTTP request may take up to 15 s, and the daemon thread
        # dies with the process anyway.
        self._thread.join(timeout=1)

    def refresh_now(self) -> None:
        """Trigger an immediate poll outside the normal schedule."""
        self._force_event.set()

    def _loop(self) -> None:
        while not self._stop_event.is_set():
            self._poll()
            timeout = self._config.poll_interval_seconds * self._backoff_factor
            self._force_event.wait(timeout=timeout)
            self._force_event.clear()

    def _poll(self) -> None:
        logger.debug("Poll cycle starting.")
        try:
            cookies = get_claude_cookies(self._config.firefox_profile)
            org_id = extract_org_id(cookies)
            cookie_header = build_cookie_header(cookies)
            data = fetch_all(
                org_id=org_id,
                cookie_header=cookie_header,
                user_agent=self._config.user_agent or None,
            )
            logger.info(
                "Poll OK: highest=%d%% tier=%s",
                data.highest_percent,
                data.subscription_tier,
            )
            self._backoff_factor = 1
            self._on_data(data)
        except RateLimitedError as exc:
            self._backoff_factor = min(self._backoff_factor * 2, 16)
            msg = str(exc)
            logger.warning(
                "Rate limited: %s (next poll in %ds)",
                msg,
                self._config.poll_interval_seconds * self._backoff_factor,
            )
            self._on_error(msg)
        except SessionExpiredError as exc:
            msg = str(exc)
            logger.warning("Session expired: %s", msg)
            self._on_error(msg)
        except (FirefoxCookieError, CookieError) as exc:
            msg = str(exc)
            logger.warning("Browser cookie error: %s", msg)
            self._on_error(msg)
        except ClaudeClientError as exc:
            msg = str(exc)
            logger.warning("API error: %s", msg)
            self._on_error(msg)
        except Exception as exc:
            msg = f"Unexpected error: {exc}"
            logger.exception("Unhandled exception in poll cycle.")
            self._on_error(msg)

File: csharp/ClaudeUsageTracker/Poller.cs
// Background polling thread.
//
// Runs ClaudeClient.FetchAll() on a configurable interval and invokes the
// registered callbacks with fresh UsageData or an error string.

namespace ClaudeUsageTracker;

public sealed class Poller
{
    private readonly Config _config;
    private readonly Action<UsageData> _onData;
    private readonly Action<string> _onError;
    private readonly ManualResetEventSlim _stopEvent = new(false);
    private readonly AutoResetEvent _forceEvent = new(false);
    private readonly Thread _thread;
    // Doubles after each consecutive 429, resets to 1 on success.
    // A manual "Refresh now" still bypasses the wait via _forceEvent.
    private int _backoffFactor = 1;

    public Poller(Config config, Action<UsageData> onData, Action<string> onError)
    {
        _config = config;
        _onData = onData;
        _onError = onError;
        _thread = new Thread(Loop) { IsBackground = true, Name = "poller" };
    }

    public void Start()
    {
        Log.Info("poller", $"Poller starting (interval={_config.PollIntervalSeconds}s)");
        _thread.Start();
    }

    public void Stop()
    {
        Log.Info("poller", "Poller stopping.");
        _stopEvent.Set();
        _forceEvent.Set();
        // Stop() can run on the UI thread (quit button): keep the join short —
        // an in-flight HTTP request may take up to 15 s, and the background
        // thread dies with the process anyway.
        _thread.Join(TimeSpan.FromSeconds(1));
    }

    /// <summary>Trigger an immediate poll outside the normal schedule.</summary>
    public void RefreshNow() => _forceEvent.Set();

    /// <summary>Doubles on each consecutive 429, capped at 16× the interval.</summary>
    internal static int NextBackoffFactor(int factor) => Math.Min(factor * 2, 16);

    private void Loop()
    {
        while (!_stopEvent.IsSet)
        {
            Poll();
            var timeout = TimeSpan.FromSeconds(_config.PollIntervalSeconds * _backoffFactor);
            _forceEvent.WaitOne(timeout);
        }
    }

    private void Poll()
    {
        Log.Debug("poller", "Poll cycle starting.");
        if (!System.Net.NetworkInformation.NetworkInterface.GetIsNetworkAvailable())
        {
            // No point hitting Cloudflare with no route — skip quietly and let
            // the widget show a network hint. The network-return trigger fires
            // an immediate poll once connectivity is back.
            Log.Debug("poller", "No network available — skipping poll.");
            _onError("Network error: no connection available.");
            return;
        }
        try
        {
            var cookies = FirefoxCookies.GetClaudeCookies(_config.FirefoxProfile);
            var orgId = FirefoxCookies.ExtractOrgId(cookies);
            var cookieHeader = FirefoxCookies.BuildCookieHeader(cookies);
            var data = ClaudeClient.FetchAll(
                orgId, cookieHeader,
                string.IsNullOrEmpty(_config.UserAgent) ? null : _config.UserAgent);
            Log.Info("poller", $"Poll OK: highest={data.HighestPercent}% tier={data.SubscriptionTier}");
            _backoffFactor = 1;
            _onData(data);
        }
        catch (RateLimitedException exc)
        {
            _backoffFactor = NextBackoffFactor(_backoffFactor);
            Log.Warning("poller",
                $"Rate limited: {exc.Message} (next poll in {_config.PollIntervalSeconds * _backoffFactor}s)");
            _onError(exc.Message);
        }
        catch (SessionExpiredException exc)
        {
            Log.Warning("poller", $"Session expired: {exc.Message}");
            _onError(exc.Message);
        }
        catch (Exception exc) when (exc is FirefoxCookieException or CookieException)
        {
            Log.Warning("poller", $"Browser cookie error: {exc.Message}");
            _onError(exc.Message);
        }
        catch (ClaudeClientException exc)
        {
            Log.Warning("poller", $"API error: {exc.Message}");
            _onError(exc.Message);
        }
        catch (Exception exc)
        {
            Log.Exception("poller", "Unhandled exception in poll cycle", exc);
            _onError($"Unexpected error: {exc.Message}");
        }
    }
}
